mycpp/gc_str.cc

```diff
@@ -462,3 +468,105 @@ List<Str*>* Str::split(Str* sep) {
 
   return result;
 }
+
+static inline Str* _StrFormat(const char* fmt, va_list args) {
+  int n = strlen(fmt);
+  auto beg = std::cregex_iterator(fmt, fmt + n, gStrFmtRegex);
+  auto end = std::cregex_iterator();
+
+  char int_buf[kMaxFmtWidth];
+  std::string buf;
+  for (std::cregex_iterator it = beg; it != end; ++it) {
+    const std::cmatch& match = *it;
+
+    const std::csub_match& lit_m = match[1];
+    assert(lit_m.matched);
+    const std::string& lit_s = lit_m.str();
+    buf.append(lit_s);
+
+    int width = 0;
+    bool zero_pad = false;
+    const std::csub_match& width_m = match[2];
+    const std::string& width_s = width_m.str();
+    if (width_m.matched && !width_s.empty()) {
+      if (width_s[0] == '0') {
+        zero_pad = true;
+        assert(width_s.size() > 1);
+        assert(StringToInteger(width_s.c_str() + 1, width_s.size() - 1, 10,
+                               &width));
+      } else {
+        assert(StringToInteger(width_s.c_str(), width_s.size(), 10, &width));
+      }
+      assert(width >= 0 && width < kMaxFmtWidth);
+    }
+
+    char const* str_to_add = nullptr;
+    const std::csub_match& code_m = match[3];
+    const std::string& code_s = code_m.str();
+    if (!code_m.matched) {
+      assert(!width_m.matched);  // python errors on invalid format operators
+      goto done;
+    }
+    assert(code_s.size() == 1);
+    switch (code_s[0]) {
+    case '%': {
+      assert(code_s.size() == 1);
+      str_to_add = code_s.c_str();
+      break;
+    }
+    case 's': {
+      Str* s = va_arg(args, Str*);
+      str_to_add = s->data();
+      zero_pad = false;  // python ignores the 0 directive for strings
+      break;
+    }
+    case 'r': {
+      Str* s = va_arg(args, Str*);
+      s = repr(s);
+      str_to_add = s->data();
+      zero_pad = false;  // python ignores the 0 directive for strings
+      break;
+    }
+    case 'd':  // fallthrough
+    case 'o': {
+      int d = va_arg(args, int);
+      assert(snprintf(int_buf, kMaxFmtWidth, match.str().c_str() + lit_s.size(),
+                      d) > 0);
+      str_to_add = int_buf;
+      break;
+    }
+    default:
+      break;
```

Comment:
I would also add an assert(0) here, even though you have one below -- it reads more nicely


```
>>> '%z' % 'foo'
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: unsupported format character 'z' (0x7a) at index 1
```